ap/inference/__init__.py

```diff
@@ -0,0 +1,1405 @@
+        """
+        # rename arguments to make their meaning explicit
+        process_name = process
+        category_pattern = category
+
+        processes = self.get_processes(
+            process_name,
+            category=category_pattern,
+            only_names=only_name,
+        )
+
+        # checks
+        if not processes:
+            if silent:
+                return None
+            err = f"no process named '{process_name}' found"
+            if category_pattern:
+                err += f" in category '{category_pattern}'"
+            raise ValueError(err)
+        if len(processes) > 1:
+            if silent:
+                return None
+            raise ValueError(
+                f"process '{process_name}' found in more than one category: " + ",".join(processes),
+            )
+
+        category, processes = list(processes.items())[0]
+        if len(processes) > 1:
+            if silent:
+                return None
+            names = processes if only_name else [p.name for p in processes]
+            raise ValueError(
+                f"process '{process_name}' found more than once in category '{category}': "
+                ",".join(names),
+            )
+
+        return processes[0]
+
+    def has_process(
+        self,
+        process: Union[str, Sequence[str]],
+        category: Optional[Union[str, Sequence[str]]] = None,
+    ) -> bool:
+        """
+        Returns *True* if a process whose name matches *process*, and optionally whose category's
+        name matches *category*, is existing, and *False* otherwise.
+        """
+        # rename arguments to make their meaning explicit
+        process_pattern = process
+        category_pattern = category
+
+        # simple length check
+        return len(self.get_processes(process_pattern, category=category_pattern)) > 0
+
+    def add_process(
+        self,
+        *args,
+        category: Optional[Union[str, Sequence[str]]] = None,
+        **kwargs,
+    ) -> None:
+        """
+        Adds a new process to all categories whose names match *category*, with all *args* and
+        *kwargs* used to create the structured process dictionary via :py:meth:`process_spec`. If a
+        process with the same name already exists in one of the categories, an exception is raised.
+        """
+        # rename arguments to make their meaning explicit
+        category_pattern = category
+
+        process = self.process_spec(*args, **kwargs)
+
+        # get categories the process should be added to
+        categories = self.get_categories(category_pattern)
+
+        # check for duplicates
+        for category in categories:
+            if self.has_process(process.name, category=category.name):
+                raise ValueError(
+                    f"process named '{process.name}' already registered in category "
+                    f"'{category.name}'",
+                )
```

Comment:
Yes and no I guess. In case misconfiguration is a risk, the exception might be helpful, but if one absolutely knows what's happening, it should indeed silently skip. I can add a flag `silent` to toggle that behavior.

Reply:
Yes, I think that would be helpful :+1: